### PR TITLE
Expose inputRef and and add required prop to allow html5 validations fixes #93

### LINKED ISSUE #93: Unable to validate input without exposing more input attributes / events
Hi, 

First of all, thanks for porting this to Svelte!

Here is a small feature request.

I regularly use the native html 5 validation api to validate my forms, but I cannot accomplish this in a satisfying way using this component. 

On a regular input I use the following attributes/events
image

I believe you should open up to add any attribute to the input. 

Hope to see this implemented.

Thanks in advance

### PR DESCRIPTION
**Pull request by zwergius**

### CODE REVIEW

**Comment by thecodejack** [comment]
@zwergius I kind of merged https://github.com/thecodejack/svelte-file-dropzone/pull/105

**Comment by zwergius** [comment]
@thecodejack ok, but #105 as far as I can see does not expose the inputRef ?

**Comment by thecodejack** [comment]
oh my bad...can you rebase and add the change?

**Comment by zwergius** [comment]
> > @thecodejack ok, but #105 as far as I can see does not expose the inputRef ?
> 
> Indeed, #105 doesn't solves this.
> 
> Notice that I have a giant PR #107 , so if you want to rebase/merge you better be waiting to the conversation there in my opinion.

@Tal500 This is actually a very simple PR - I hadn't got around to clean it up yet, most of the changes are prettier which probably should not be there... So the only real addition here is 2 new props: inputRef & required, which are both added to the input. You change being so big, I'd probably like to get this in before. It should be very straightforward to merge....

File: src/components/Dropzone.svelte
Comment (Tal500) [inline src/components/Dropzone.svelte:217]: No need to have "?" at `event.dataTransfer?`.

**Comment by zwergius** [comment]
@Tal500 @thecodejack Just rebased and removed formatted changes + added props in README

File: README.md
Comment (Tal500) [inline README.md:66]: Details are on the following comments.
```suggestion
| inputElement              | reference to inputRef                                                                    | undefined     |

```

File: src/components/Dropzone.svelte
Comment (Tal500) [inline src/components/Dropzone.svelte:35]: If exposing this var, we should make this var have a good name.
I think that `inputElement` is the best name because the type is `HTMLInputElement`.

```suggestion
  export let inputElement;
```

File: package.json
Comment (Tal500) [inline package.json:11]: I think that it's better to remove this line, since you don't want to force the user to build right away the package on install.
```suggestion
```

Comment (zwergius) [reply package.json:11]: I added this to use this package straight from my GH, so yea we can remove it

File: src/components/Dropzone.svelte
Comment (zwergius) [reply src/components/Dropzone.svelte:217]: its already inside ```if (isEvtWithFiles(event))```

Comment (zwergius) [reply src/components/Dropzone.svelte:217]: its already inside if (isEvtWithFiles(event))

Comment (Tal500) [reply src/components/Dropzone.svelte:217]: Not exactly, it checks in `isEvtWithFiles` if files where received,

What I mean is
```suggestion
          inputRef.files = event.dataTransfer.files;

```
since the `if (event.dataTransfer) {` already promises that `event.dataTransfer` isn't `null` or `undefined`.

File: src/components/Dropzone.svelte
Comment (Tal500) [inline src/components/Dropzone.svelte:35]: Thanks for changing its name, but notice you need to change all of the other references of `inputRef` in the code, to become `inputElement`.
To be sure, I suggest to check if it gets compiled well after renaming.

File: src/components/Dropzone.svelte
Comment (zwergius) [reply src/components/Dropzone.svelte]: ah right, the other way around - pretty sure I added this for some reason...

File: src/components/Dropzone.svelte
Comment (zwergius) [reply src/components/Dropzone.svelte:35]: hehe yea that went a bit quick

File: src/components/Dropzone.svelte
Comment (zwergius) [reply src/components/Dropzone.svelte]: removed 👍

**Review by Tal500 — APPROVED** [review]

**Comment by Tal500** [comment]
One thing that I am afraid about is that when you export the variable `inputElement`, you allow it to be modified by the user (Svelte doesn't have readonly props, they are all bidirectional).

Since the value of this var is set after mount and stay fixed until component destruction, I suggest the following:
Hide the variable `inputElement`, but export a function `getInputElement` (inside the component).
This function returns `inputElement` if the component where already mounted (which happens whenever `inputRef != undefined`), and throw an exception otherwise.
I think it's the safest and clearest solution. No need to worry for varaible reactivity, since the input element stays constant after mounting, before destruction.

**Comment by zwergius** [comment]
> One thing that I am now afraid about is that when you export the variable `inputElement`, you allow it to be modified by the user (Svelte doesn't have readonly props, they are all bidirectional).
> 
> Since the value of this var is set after mount and stay fixed until component destruction, I suggest the following: Hide the variable `inputElement`, but export a function `getInputElement` (inside the component). This function returns `inputElement` if the component where already mounted (which happens whenever `inputRef != undefined`), and throw an exception otherwise. I think it's the safest and clearest solution. No need to worry for varaible reactivity, since the input element stays constant after mounting, before destruction.

I really don't see that as an issue, this is an optional prop that you set if you need to interact with the inputElement, just as you would with any other element, if you know you need the element, then I reckon you also know not to modify it . Exporting a function isn't very svelte like...

**Comment by thecodejack** [comment]
I think what @Tal500 referring to is "what will happen if user updates a value the the `inputElement`?" 

So @Tal500 I am guessing we are binding inputElement with component rather setting value to inputElement. I am guessing it won't be a problem but worth checking before merge.

If possible @zwergius can you quick confirm that won't be an issue?

**Comment by zwergius** [comment]
@thecodejack You mean like setting the file from outside the component ? You'd have to create another input and through file uploaded there set this input.. Not sure I see that as something you need to worry about...

**Comment by Tal500** [comment]
I thought about some good compromise. 
Change the var to be the internal variable `_inputElement`, do not expose it, and use it(and only it).
Then, define the following lines of code:
```js
$: inputElement = _inputElement;
export { inputElement };
```

This means that the user can bind to the public (reactive) varaible `inputElement`.
If the user decides or by mistake change the variable value, it's his fault and it won't affect the internal logic of the library.

**Comment by zwergius** [comment]
> I thought about some good compromise. Change the var to be the internal variable `_inputElement`, do not expose it, and use it(and only it). Then, define the following lines of code:
> 
> ```js
> $: inputElement = _inputElement;
> export { inputElement };
> ```
> 
> This means that the user can bind to the public (reactive) varaible `inputElement`. If the user decides or by mistake change the variable value, it's his fault and it won't affect the internal logic of the library.

I don't understand why this is such a huge issue, but if you insist on this maybe it's simpler if you just push this change yourself? I only added this in order to tap into some input events that aren't exposed so I can use the native form validation...   I could have achieved the same by adding callbacks for the missing events but in the end I thought that this component would be more flexible if exposing the inputElement...

**Comment by Tal500** [comment]
> > I thought about some good compromise. Change the var to be the internal variable `_inputElement`, do not expose it, and use it(and only it). Then, define the following lines of code:
> > ```js
> > $: inputElement = _inputElement;
> > export { inputElement };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This means that the user can bind to the public (reactive) varaible `inputElement`. If the user decides or by mistake change the variable value, it's his fault and it won't affect the internal logic of the library.
> 
> I don't understand why this is such a huge issue, but if you insist on this maybe it's simpler if you just push this change yourself? I only added this in order to tap into some input events that aren't exposed so I can use the native form validation... I could have achieved the same by adding callbacks for the missing events but in the end I thought that this component would be more flexible if exposing the inputElement...

Just wanted to see the maintainer thoughts before implementing.

**Comment by zwergius** [comment]
Any plans to get this merged at some point ?

**Comment by easteryard** [comment]
Do you have an update on this @thecodejack? Got a project which depends on it and would really like to know if we can expect it to be merged any time soon.

**Comment by thecodejack** [comment]
I have merged this. Will publish soon

**Pull request merged**
